### PR TITLE
chore(examples/templates/aws-devcontainer): update to use envbuilder provider`

### PR DESCRIPTION
**Pull request by johnstcn**
Relates to https://github.com/coder/coder/issues/13278

- Updates to use envbuilder provider
- Allows adding a public SSH key for troubleshooting
- Adds support for attaching an instance profile (for registries hosted on ECR)
- Also fixes an issue in the gcp-devcontainer template where we do not actually use the output of `cached_image` 💀 

You can play around with the current version of the template here (internal link): https://dev.coder.com/templates/coder/cian-aws-devcontainer

### CODE REVIEW

File: examples/templates/aws-devcontainer/main.tf [hunk 1]
```diff
@@ -144,24 +216,57 @@ locals {
...
 
+  # Set up Docker credentials
+  mkdir -p "/home/${local.linux_user}/.docker"
+  if [ -n "${local.docker_config_json_base64}" ]; then
+     # Write the Docker config JSON to disk if it is provided.
+     printf "%s" "${local.docker_config_json_base64}" | base64 -d | tee "/home/${local.linux_user}/.docker/config.json"
+  else
+    # Assume that we're going to use the instance IAM role to pull from the cache repo if we need to.
+    # Set up the ecr credential helper.
+    apt-get update -y && apt-get install -y amazon-ecr-credential-helper
```
Comment (matifali) [inline examples/templates/aws-devcontainer/main.tf:227]: Does it work without configuring a repo? I mean is it available in default Ubuntu repos?

File: examples/templates/aws-devcontainer/main.tf [hunk 2]
```diff
@@ -144,24 +216,57 @@ locals {
...
 resource "aws_instance" "vm" {
-  ami               = data.aws_ami.ubuntu.id
-  availability_zone = "${module.aws_region.value}a"
-  instance_type     = data.coder_parameter.instance_type.value
+  ami                  = data.aws_ami.ubuntu.id
+  availability_zone    = local.aws_availability_zone
+  instance_type        = data.coder_parameter.instance_type.value
+  iam_instance_profile = try(data.aws_iam_instance_profile.vm_instance_profile[0].name, null)
   root_block_device {
     volume_size = 30
```
Comment (matifali) [inline examples/templates/aws-devcontainer/main.tf:271]: Should we also make the root disk size a parameter?

File: examples/templates/aws-devcontainer/main.tf [hunk 3]
```diff
@@ -181,3 +286,62 @@ resource "aws_ec2_instance_state" "vm" {
...
+  }
+  item {
+    key   = "devcontainer_builder"
+    value = data.coder_parameter.devcontainer_builder.value
+  }
+}
+
+module "code-server" {
+  count    = data.coder_workspace.me.start_count
+  source   = "https://registry.coder.com/modules/code-server"
```
Comment (matifali) [inline examples/templates/aws-devcontainer/main.tf:345]: version is required otherwise it will default to v1.0.0.

**Review by matifali — APPROVED** [review]
Great job. ❤️

File: examples/templates/aws-devcontainer/main.tf (hunk 1 above)
Comment (johnstcn) [reply examples/templates/aws-devcontainer/main.tf:227]: yep

File: examples/templates/aws-devcontainer/main.tf (hunk 2 above)
Comment (johnstcn) [reply examples/templates/aws-devcontainer/main.tf:271]: I'd like to keep this a separate change.

**Pull request merged**